File: golang/802.find-eventual-safe-states.go
func eventualSafeNodes(graph [][]int) []int {
	N := len(graph)
	revGraph := make([][]int, N)
	outDegree := make([]int, N)
	safeNodes := make([]bool, N)
	safe := make([]int, 0)
	que := make([]int, 0, N)

	for i := 0; i < N; i++ {
		for _, j := range graph[i] {
			revGraph[j] = append(revGraph[j], i)
			outDegree[i]++
		}

		if len(graph[i]) == 0 {
			que = append(que, i)
		}
	}

	for len(que) > 0 {
		node := que[0]
		que = que[1:]
		safeNodes[node] = true

		for _, nNode := range revGraph[node] {
			outDegree[nNode]--
			if outDegree[nNode] == 0 {
				que = append(que, nNode)
			}
		}
	}

	for i := 0; i < N; i++ {
		if safeNodes[i] {
			safe = append(safe, i)
		}
	}

	return safe
}

File: python/802.find-eventual-safe-states.py
from collections import deque


class Solution:
    def eventualSafeNodes(self, graph: list[list[int]]) -> list[int]:
        n = len(graph)
        revEdges = [[] for _ in range(n)]
        safeNodes = [False] * n
        outDegres = [0] * n

        for u, nodes in enumerate(graph):
            for v in nodes:
                revEdges[v].append(u)
                outDegres[u] += 1

        queue = deque([i for i in range(n) if outDegres[i] == 0])

        while queue:
            node = queue.popleft()
            safeNodes[node] = True

            for nNode in revEdges[node]:
                outDegres[nNode] -= 1
                if outDegres[nNode] == 0:
                    queue.append(nNode)

        return [node for node, safety in enumerate(safeNodes) if safety]

